# 🧪 Pytest Complete Tutorial

This notebook is a **complete tutorial for pytest**, including:
- Basic setup
- Writing test functions
- Assertions
- Fixtures
- Parametrization
- Mocking
- Running pytest in Jupyter
- Best practices

## 1. Installing pytest

You can install pytest via pip:

In [ ]:
!pip install pytest

## 2. Writing Your First Test

Pytest discovers functions prefixed with `test_` in files named `test_*.py` or `*_test.py`. Let's create a simple function and test it.

In [ ]:
def add(a, b):
    return a + b

def test_add():
    assert add(2, 3) == 5
    assert add(-1, 1) == 0
    assert add(0, 0) == 0

## 3. Running Tests in Jupyter Notebook

We can save the code to a file and run pytest:

In [ ]:
with open('test_example.py', 'w') as f:
    f.write('''
def add(a, b):
    return a + b

def test_add():
    assert add(2, 3) == 5
    assert add(-1, 1) == 0
    assert add(0, 0) == 0
''')

!pytest -v test_example.py

============================= test session starts =============================
collected 1 item                                                                

test_example.py::test_add PASSED                                          [100%]

============================== 1 passed in 0.02s =============================


## 4. Assertions

Pytest uses simple Python `assert` statements.

In [1]:
def test_assertions():
    # Equality
    assert 2 + 2 == 4
    # Boolean
    assert True
    # Membership
    assert 'Py' in 'Pytest'
    # Raises Exception
    import math
    try:
        math.sqrt(-1)
    except ValueError as e:
        assert str(e) == 'math domain error'

In [3]:
with open('Assertions_test_example.py', 'w') as f:
    f.write('''
def test_assertions():
    # Equality
    assert 2 + 2 == 4
    # Boolean
    assert True
    # Membership
    assert 'Py' in 'Pytest'
    # Raises Exception
    import math
    try:
        math.sqrt(-1)
    except ValueError as e:
        assert str(e) == 'math domain error'
''')

!pytest -v Assertions_test_example.py

============================= test session starts ==============================
platform linux -- Python 3.12.12, pytest-8.4.2, pluggy-1.6.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /content
plugins: typeguard-4.4.4, langsmith-0.4.59, anyio-4.12.0
collected 1 item                                                               

Assertions_test_example.py::test_assertions PASSED                       [100%]

============================== 1 passed in 0.02s ===============================


## 5. Fixtures

Fixtures help set up and clean up resources for tests.

In [5]:
import pytest

@pytest.fixture
def sample_list():
    return [1, 2, 3]

def test_fixture(sample_list):
    assert sum(sample_list) == 6
    sample_list.append(4)
    assert sample_list[-1] == 4

In [6]:
with open('Fixture_test_example.py', 'w') as f:
    f.write('''
import pytest

@pytest.fixture
def sample_list():
    return [1, 2, 3]

def test_fixture(sample_list):
    assert sum(sample_list) == 6
    sample_list.append(4)
    assert sample_list[-1] == 4
''')

!pytest -v Fixture_test_example.py

============================= test session starts ==============================
platform linux -- Python 3.12.12, pytest-8.4.2, pluggy-1.6.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /content
plugins: typeguard-4.4.4, langsmith-0.4.59, anyio-4.12.0
collected 1 item                                                               

Fixture_test_example.py::test_fixture PASSED                             [100%]

============================== 1 passed in 0.01s ===============================


## 6. Parametrized Tests

Run the same test with multiple inputs using `@pytest.mark.parametrize`.

In [7]:
@pytest.mark.parametrize("a,b,expected", [
    (2, 3, 5),
    (-1, 1, 0),
    (0, 0, 0)
])
def test_add_param(a, b, expected):
    assert add(a, b) == expected

In [8]:
with open('Parametrized_test_example.py', 'w') as f:
    f.write('''
import pytest

@pytest.fixture
def sample_list():
    return [1, 2, 3]

def test_fixture(sample_list):
    assert sum(sample_list) == 6
    sample_list.append(4)
    assert sample_list[-1] == 4
''')

!pytest -v Parametrized_test_example.py

============================= test session starts ==============================
platform linux -- Python 3.12.12, pytest-8.4.2, pluggy-1.6.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /content
plugins: typeguard-4.4.4, langsmith-0.4.59, anyio-4.12.0
collected 1 item                                                               

Parametrized_test_example.py::test_fixture PASSED                        [100%]

============================== 1 passed in 0.01s ===============================


## 7. Testing Exceptions

Use `pytest.raises` to assert exceptions.

In [10]:
def divide(a, b):
    return a / b

def test_divide_zero():
    with pytest.raises(ZeroDivisionError):
        divide(1, 0)

In [9]:
with open('Exceptions_test_example.py', 'w') as f:
    f.write('''
import pytest

@pytest.fixture
def sample_list():
    return [1, 2, 3]

def test_fixture(sample_list):
    assert sum(sample_list) == 6
    sample_list.append(4)
    assert sample_list[-1] == 4
''')

!pytest -v Exceptions_test_example.py

============================= test session starts ==============================
platform linux -- Python 3.12.12, pytest-8.4.2, pluggy-1.6.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /content
plugins: typeguard-4.4.4, langsmith-0.4.59, anyio-4.12.0
collected 1 item                                                               

Exceptions_test_example.py::test_fixture PASSED                          [100%]

============================== 1 passed in 0.02s ===============================


## 8. Skipping and Expected Failures

You can skip tests or mark expected failures.

In [12]:
@pytest.mark.skip(reason="Not implemented yet")
def test_skip():
    assert False

@pytest.mark.xfail(reason="This test will fail")
def test_expected_fail():
    assert 1 == 2

In [11]:
with open('Skipping_and_Expected_test_example.py', 'w') as f:
    f.write('''
@pytest.mark.skip(reason="Not implemented yet")
def test_skip():
    assert False

@pytest.mark.xfail(reason="This test will fail")
def test_expected_fail():
    assert 1 == 2
''')

!pytest -v Exceptions_test_example.py

============================= test session starts ==============================
platform linux -- Python 3.12.12, pytest-8.4.2, pluggy-1.6.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /content
plugins: typeguard-4.4.4, langsmith-0.4.59, anyio-4.12.0
collected 1 item                                                               

Exceptions_test_example.py::test_fixture PASSED                          [100%]

============================== 1 passed in 0.01s ===============================


## 9. Mocking

Use `unittest.mock` to replace parts of your system during tests.

In [15]:
from unittest.mock import Mock

def test_mock():
    mock_obj = Mock()
    mock_obj.method.return_value = 42
    assert mock_obj.method() == 42
    mock_obj.method.assert_called_once()

In [17]:
with open('Mocking_test_example.py', 'w') as f:
    f.write('''
from unittest.mock import Mock

def test_mock():
    mock_obj = Mock()
    mock_obj.method.return_value = 42
    assert mock_obj.method() == 42
    mock_obj.method.assert_called_once()
''')

!pytest -v Mocking_test_example.py

============================= test session starts ==============================
platform linux -- Python 3.12.12, pytest-8.4.2, pluggy-1.6.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /content
plugins: typeguard-4.4.4, langsmith-0.4.59, anyio-4.12.0
collected 1 item                                                               

Mocking_test_example.py::test_mock PASSED                                [100%]

============================== 1 passed in 0.01s ===============================


## 10. Best Practices

- Keep test functions small and focused.
- Use fixtures for repeated setup.
- Name test files `test_*.py`.
- Use `pytest.mark.parametrize` for multiple inputs.
- Use `pytest.raises` for exceptions.
- Keep tests independent of each other.
- Use mocking for external dependencies.

## 11. Running Tests Options

- Run all tests: `pytest`
- Verbose mode: `pytest -v`
- Run a specific test file: `pytest test_example.py`
- Run a specific test function: `pytest -k test_add`
- Show print statements: `pytest -s`

🎉 **Congratulations!** You now have a full, productive `pytest` tutorial in a single notebook compatible with Python 3.12.